server/src/main/java/org/opensearch/search/startree/filter/provider/StarDateFieldMapper.java

```diff
@@ -0,0 +1,152 @@
+            ? DateFormatter.forPattern(rangeQueryBuilder.format()).toDateMathParser()
+            : org.opensearch.index.mapper.DateFieldMapper.getDefaultDateTimeFormatter().toDateMathParser();
+
+        ZoneId timeZone = rangeQueryBuilder.timeZone() != null ? ZoneId.of(rangeQueryBuilder.timeZone()) : null;
+
+        long l = Long.MIN_VALUE;
+        long u = Long.MAX_VALUE;
+
+        if (rangeQueryBuilder.from() != null) {
+            l = DateFieldType.parseToLong(
+                rangeQueryBuilder.from(),
+                !rangeQueryBuilder.includeLower(),
+                timeZone,
+                forcedDateParser,
+                nowSupplier,
+                dateFieldType.resolution()
+            );
+            if (!rangeQueryBuilder.includeLower()) {
+                ++l;
+            }
+        }
+
+        if (rangeQueryBuilder.to() != null) {
+            u = DateFieldType.parseToLong(
+                rangeQueryBuilder.to(),
+                rangeQueryBuilder.includeUpper(),
+                timeZone,
+                forcedDateParser,
+                nowSupplier,
+                dateFieldType.resolution()
+            );
+            if (!rangeQueryBuilder.includeUpper()) {
+                --u;
+            }
+        }
+
+        // Find the matching interval - preferring the highest possible interval for query optimization
+        List<DateTimeUnitRounding> intervals = dateDimension.getSortedCalendarIntervals().reversed();
+        DateTimeUnitRounding matchingInterval = null;
+
+        for (DateTimeUnitRounding interval : intervals) {
+            long roundedLow = l != Long.MIN_VALUE ? interval.roundFloor(l) : l;
+            long roundedHigh = u != Long.MAX_VALUE ? interval.roundFloor(u) : u;
+
+            // This is needed since OpenSearch rounds up to the last millisecond in the rounding interval.
+            // so when the date parser rounds up to say 2022-05-31T23:59:59.999 we can check if by adding 1
+            // the new interval which is 2022-05-31T00:00:00.000 in the example can be solved via star tree
+            //
+            // this is not needed for low since rounding is on the first millisecond 2022-06-01T00:00:00.000
+            long roundedHighPlus1 = u != Long.MAX_VALUE ? interval.roundFloor(u + 1) : u;
+
+            // If both bounds round to the same values, we have an exact match for this interval
+            if (roundedLow == l && (roundedHigh == u || roundedHighPlus1 == u + 1)) {
+                matchingInterval = interval;
+                break; // Found the most granular matching interval
+            }
+        }
+
+        if (matchingInterval == null) {
+            return null; // No matching interval found, fall back to default implementation
+        }
+
+        // Construct the sub-dimension field name
+        subDimensionField = field + "_" + matchingInterval.shortName();
+
+        return new RangeMatchDimFilter(
```

Comment:
So since the timestamp is rounded off, hour 1 stores all the values between 1:00 and 1:59 and hour 2 stores all the values between 2:00 and 2:59.

For example

Say my query is between 1:00 and 2:00

1st hour covers 1:00 to 1:59
2nd hour covers 2:00 to 2:59

aren't we querying for range of 1:00 and 2:00 including L and U, as part of range dim filter , thereby querying 1:00 to 2:59 instead of 1:00 to 1:59 ?

Reply:
If this is true, this is a bug and its not captured in tests. Lets add cases with actual dates in `DateRangeTests` which validates such scenarios.

When we do random - we might not end up with test cases which falls exactly in boundary conditions like above.

Reply:
Let me help you with an example of a valid query:

```
gte": "2024-01-24T00:00:00.000Z",
"lt": "2024-02-26T00:00:00.000Z"
```

The converted long values (epochMillis) are as follows:

```
"gte": 1706054400000
"lt": 1708905600000
```

Dates irrespective of their timezone/format are converted to these long values. We deal with star-tree traversing with these values only.

This `l` and `u` passed in the `RangeMatchDimFilter` initialization are: `1706054400000` & `1708905599999` respectively, with `includeLow` and `includeHigh` as true.

This is effective as writing the query as:

```
"gte": 1706054400000
"lte": 1708905599999
```
which is equivalent to:

```
gte": "2024-01-24T00:00:00.000Z",
"lte": "2024-02-25T23:59:59.999Z"
```

Reply:
This whole [example](https://github.com/opensearch-project/OpenSearch/pull/17855#discussion_r2048299783) applies for lte,gte. With milliseconds. 

The bug still exists but the mentioned range was wrong.